### PR TITLE
AMSUA bugfix

### PR DESCRIPTION
**Pull request by timhoar**
This addresses Issue 208.  The AMSUA converter looped over multiple input files but only wrote out observations for the last input file. This corrects that error.

Since there are hundreds of files per day for this instrument, appending to an existing file is supported.
Because there are hundreds of files per day, a verbosity flag was added.

### CODE REVIEW

File: observations/obs_converters/AIRS/amsua_netCDF_support_mod.f90
Comment (nancycollins) [inline observations/obs_converters/AIRS/amsua_netCDF_support_mod.f90:1134]: i'm a little concerned about the performance of this approach.  as the output files get larger the cost of inserting an obs in the file grows.  the original code added new obs to the eventual output sequence.  this version creates a separate sequence (one insert here) and then combines them (another insert).  could this code pass the eventual output sequence into the fill routine and have it put the obs where they are eventually going to be?

File: observations/obs_converters/AIRS/convert_amsu_L1.f90
Comment (nancycollins) [inline observations/obs_converters/AIRS/convert_amsu_L1.f90:141]: the original main execution loop in this code was very clean.  could this code (~lines 139-157) be moved to a subroutine?  putting if statements into the main section of code obscures the high level logic.

File: observations/obs_converters/AIRS/convert_amsu_L1.f90
Comment (nancycollins) [inline observations/obs_converters/AIRS/convert_amsu_L1.f90:168]: if the create step is moved out of the loop and obs are added to the same sequence then you don't have to combine them as a separate step.  if there's a reason for having to do it this way there needs to be some comment before this loop saying why this choice was made.  it is going to have worse performance for large obs counts.

File: observations/obs_converters/AIRS/convert_amsu_L1.f90
Comment (nancycollins) [inline observations/obs_converters/AIRS/convert_amsu_L1.f90:193]: i'm not sure i'd make this choice to error out if no obs are there.  in a big batch conversion, it's possible there are missing files, etc.  printing a msg is great but an error stops things abruptly.

**Review by nancycollins — COMMENTED** [review]
most of the changes look great.  i do have a concern about adding more conditional code (if statements) to the original simple main logic loop.  also the process of moving obs from one sequence to another is going to cost more in execution time than adding the obs directly to the output sequence.

File: observations/obs_converters/AIRS/amsua_netCDF_support_mod.f90
Comment (timhoar) [reply observations/obs_converters/AIRS/amsua_netCDF_support_mod.f90:1134]: I understand ... mostly.  The separate obs sequence (the one for each file) is essentially guaranteed to be an append for each observation - not a searching insert. That should be about as fast as possible.  By creating a separate sequence for each file, it is possible to report on the full characteristics of THAT sequence, as opposed to the incremental change to the output sequence. Maybe that's not worthwhile ... dunno. 
The combination insert is where the inefficiency (if any) lies. If the user (and that's a big if) puts the input files in temporally ascending order, these are all appends - again - about as efficient as possible. If the files listed are out-of-order, they are their own worst enemy to begin with. 
The logic of the code got a bit confusing when I tried to use a single sequence, and had a pretty large footprint on the code.

File: observations/obs_converters/AIRS/convert_amsu_L1.f90
Comment (timhoar) [reply observations/obs_converters/AIRS/convert_amsu_L1.f90:141]: Agreed ... I was going to do that after it passed, and have forgotten. Thanks for the reminder.

File: observations/obs_converters/AIRS/convert_amsu_L1.f90
Comment (timhoar) [reply observations/obs_converters/AIRS/convert_amsu_L1.f90:168]: There are about 240 files per day, and using two channels from each file results in 2700 obs per file, 648000 obs per day.
With gfortran -O2 optimization on my laptop, it took less than 30 seconds to convert all 240 files in a single run - and that was using the (obs_def_rttov_mod:MAXrttovkey) initial metadata length of 100,000.

I guess I just felt the code was pretty clear by adding the extra routine and leaving the existing working code as untouched as possible.

File: observations/obs_converters/AIRS/convert_amsu_L1.f90
Comment (timhoar) [reply observations/obs_converters/AIRS/convert_amsu_L1.f90:193]: These filenames are pretty unpredictable so the only way I know of to create them is to capture a listing - so I'm not really concerned about missing files ... now ... files with no valid observations is another thing. I haven't run into that and am not sure what the intent is.  I'm willing to let the end user modify this to a message if they want. I don't want the default behavior to be to just blow by 'empty/bad' files - particularly in a batch job where no one actually looks at the output.

File: observations/obs_converters/AIRS/amsua_netCDF_support_mod.f90
Comment (nancycollins) [reply observations/obs_converters/AIRS/amsua_netCDF_support_mod.f90:1134]: i might assume that someone who wanted to print detailed output about the input files might use a different program (obs_info, for example).  printing 240 detailed summaries in a single output file seems hard to find what you are looking for.   but i can see the logic in making separate sequences and concatinating them.  so as long as the performance is ok then this seems ok to me.

File: observations/obs_converters/AIRS/convert_amsu_L1.f90
Comment (nancycollins) [reply observations/obs_converters/AIRS/convert_amsu_L1.f90:168]: i don't know that leaving the existing code untouched is justification for this, but i am convinced by the performance argument.   someone doing a long assimilation may be converting months (years) of these obs, but a run time of 30 seconds a day seems like it would be acceptable in a batch job.

File: observations/obs_converters/AIRS/convert_amsu_L1.f90
Comment (nancycollins) [reply observations/obs_converters/AIRS/convert_amsu_L1.f90:193]: in batch jobs i actually always grepped the output logs because real world obs files are such a crapshoot.  i don't agree with making this an error because if i've waited in the batch queue for an hour and i'm converting 30 days of data and the first one has a problem i get nothing for my trouble.  but like you said, the user can always modify the code.  i'd let this go if you add a comment above the error line saying something like "change E_ERR to E_MSG to allow this program to exit without returning an error code."

**Comment by nancycollins** [comment]
this is just for the record - not asking for any changes here.

the reason i look at efficiency in obs converters is because of history.  several of the madis obs converters starting taking upwards of 5 or 10 *minutes* to create output files.  the culprit was the large number of obs and them not being in time order in the input netcdf files.  my solution was to index sort them by time order after reading them in but before adding them to the output obs_seq.  doing that made the run time drop back down to something reasonable.   the NCEP BUFR converter also will run pretty slow if the obs counts get high.  BUFR obs are never in time order.   you have to call a BUFR read routine to get the next obs so there is never a time when you have them all available to do a time sort on them.  small output files don't have this problem, it only shows up when the obs counts get large.  you are correct that combining sequences is fast if both sequences are already sorted in time order.   if something related to creating obs files starts to get slow, look at obs times and order of insertion.  it is true that the code appends new obs to the end of the physical file so that existing obs don't have to be copied to insert a new obs.  but the linked list still has to be searched to know where to insert the new obs in the list to maintain the logical time order.  that search can get slow.

File: observations/obs_converters/AIRS/convert_amsu_L1.f90
Comment (timhoar) [reply observations/obs_converters/AIRS/convert_amsu_L1.f90:168]: As another data point, I 'appended' the exact same 240 files to itself, which has to be a rather awful insertion pattern. The run-time (only did it once, so take it with a grain of salt) ... was 57 seconds.  Seemed to me that it took a bit of time to read the existing sequence, and write the 2X sequence ... as opposed to reading and inserting the new obs. for what its worth ...

File: observations/obs_converters/AIRS/amsua_netCDF_support_mod.f90
Comment (timhoar) [inline observations/obs_converters/AIRS/amsua_netCDF_support_mod.f90:142]: FYI only: this particular check makes it impossible to use a netCDF file with a single pixel as test input

File: observations/obs_converters/AIRS/convert_amsu_L1.f90
Comment (timhoar) [inline observations/obs_converters/AIRS/convert_amsu_L1.f90:179]: FYI: when the error_handler exits with an error in this case, the program status is 99.
When it exits with E_MSG, the program status is 0.

**Review by timhoar — COMMENTED** [review]
I think it is ready for testing.
Just so people know ... I committed version 2da9c972f9  which writes out exactly 1 observation per file. I used the following to extract the same observations (all channels, however) so it is possible to confirm that appending,etc. is being done correctly. The debug version can be used in conjunction with the output from these commands:
```
#!/bin/csh
@ fnum = 1
foreach FILE ( AIRS.2019.06.22.23?.*.nc )
   ncks -O -d GeoTrack,10 -d GeoXTrack,20 $FILE output_${fnum}.nc
   ncdump -v brightness_temp output_${fnum}.nc | tail -n 5
   @ fnum ++
end
```
I have put some test datafiles in _/glade/work/thoar/git/DART_public/observations/obs_converters/AIRS/data_

File: observations/obs_converters/AIRS/convert_amsu_L1.f90
Comment (timhoar) [reply observations/obs_converters/AIRS/convert_amsu_L1.f90]: Done ... the routine is called **append_or_create()** and is in the amsua_netCDF_support_mod.

File: observations/obs_converters/AIRS/convert_amsu_L1.f90
Comment (timhoar) [reply observations/obs_converters/AIRS/convert_amsu_L1.f90]: I added comments and made the output string mention the comment in the code, so they know which one to edit.

File: observations/obs_converters/AIRS/convert_amsu_L1.f90
Comment (nancycollins) [reply observations/obs_converters/AIRS/convert_amsu_L1.f90:179]: yes, that's exactly what i would expect.

**Review by nancycollins — APPROVED** [review]
was there a doc file for this converter?  if so, the updated namelist item needs to be added.  but i know there are several converters w/o doc files, so i'm approving this pull request even if there isn't one. (maybe we should be adding skeletal doc files saying 'documentation TBD' or something? so users know they shouldn't keep looking for a non-existent file?)

**Comment by timhoar** [comment]
> was there a doc file for this converter? if so, the updated namelist item needs to be added. but i know there are several converters w/o doc files, so i'm approving this pull request even if there isn't one. (maybe we should be adding skeletal doc files saying 'documentation TBD' or something? so users know they shouldn't keep looking for a non-existent file?)

Actually, there is a _README.rst_ and an _AIRS.rst_ in the AIRS directory. Consolidating these is already an issue - it is confusing to have two. The _AIRS.rst_ is the converted contents of the AIRS.html, which does not have any information about the `convert_amsu_L1.f90` program. The _README.rst_ does not have any namelist information in it at all. Consolidating these two needs to be done - soon - and would be the right time to document the _convert_amsu_L1_ namelist. Should this be part of the PR?

**Comment by hkershaw-brown** [comment]
> > was there a doc file for this converter? if so, the updated namelist item needs to be added. but i know there are several converters w/o doc files, so i'm approving this pull request even if there isn't one. (maybe we should be adding skeletal doc files saying 'documentation TBD' or something? so users know they shouldn't keep looking for a non-existent file?)
> 
> Actually, there is a _README.rst_ and an _AIRS.rst_ in the AIRS directory. Consolidating these is already an issue - it is confusing to have two. The _AIRS.rst_ is the converted contents of the AIRS.html, which does not have any information about the `convert_amsu_L1.f90` program. The _README.rst_ does not have any namelist information in it at all. Consolidating these two needs to be done - soon - and would be the right time to document the _convert_amsu_L1_ namelist. Should this be part of the PR?

I think it is a really good time to update/correct the documentation.  It seems well within the scope: fix the converter, documentation fix for people to use the converter correctly.

**Comment by timhoar** [comment]
> > > was there a doc file for this converter? if so, the updated namelist item needs to be added. but i know there are several converters w/o doc files, so i'm approving this pull request even if there isn't one. (maybe we should be adding skeletal doc files saying 'documentation TBD' or something? so users know they shouldn't keep looking for a non-existent file?)
> > 
> > 
> > Actually, there is a _README.rst_ and an _AIRS.rst_ in the AIRS directory. Consolidating these is already an issue - it is confusing to have two. The _AIRS.rst_ is the converted contents of the AIRS.html, which does not have any information about the `convert_amsu_L1.f90` program. The _README.rst_ does not have any namelist information in it at all. Consolidating these two needs to be done - soon - and would be the right time to document the _convert_amsu_L1_ namelist. Should this be part of the PR?
> 
> I think it is a really good time to update/correct the documentation. It seems well within the scope of this PR: fix the converter, documentation fix for people to use the converter correctly.

OK - I'm on it ... combining AIRS.rst into  README.rst  ... please do not merge the pull request yet!

File: observations/obs_converters/AIRS/work/input.nml
Comment (nancycollins) [inline observations/obs_converters/AIRS/work/input.nml]: is this still an appropriate filename?  and is this file included or is it just an example?

**Review by nancycollins — APPROVED** [review]
all the docs look great.  i had one question about the namelist but it's not that big a deal, i'm just curious.  hurrah.

File: observations/obs_converters/AIRS/work/input.nml
Comment (nancycollins) [inline observations/obs_converters/AIRS/work/input.nml:53]: this looks good.

**Review by nancycollins — APPROVED** [review]

**Pull request merged**
